File: cmd/igor/internal/template/funcmap.go
package template

import (
	"strings"
	"text/template"

	"github.com/joshjhall/containers/cmd/igor/internal/feature"
)

// BuildArgGroup represents a category of build args with a label header.
type BuildArgGroup struct {
	Label string          // "Languages", "Tools", etc.
	Args  []BuildArgEntry // ordered entries within the group
}

// BuildArgEntry is a single build arg line with an optional inline comment.
type BuildArgEntry struct {
	Line    string // e.g. "INCLUDE_PYTHON_DEV=true"
	Comment string // e.g. "includes Python runtime" (empty if none)
}

// categoryLabel maps feature categories to display labels.
var categoryLabel = map[feature.Category]string{
	feature.CategoryLanguage: "Languages",
	feature.CategoryTool:     "Tools",
	feature.CategoryAI:       "AI/ML",
	feature.CategoryCloud:    "Cloud",
	feature.CategoryDatabase: "Database",
}

// categoryOrder defines the rendering order for categories.
var categoryOrder = []feature.Category{
	feature.CategoryLanguage,
	feature.CategoryTool,
	feature.CategoryAI,
	feature.CategoryCloud,
	feature.CategoryDatabase,
}

// groupedBuildArgs groups enabled features by category, collapses base+dev
// pairs (emitting only the dev line with an annotation), and attaches version
// args after the corresponding feature entry.
func groupedBuildArgs(ctx *RenderContext) []BuildArgGroup {
	// Build a set of enabled feature IDs for quick lookup.
	enabledIDs := make(map[string]bool, len(ctx.EnabledFeatures))
	for _, f := range ctx.EnabledFeatures {
		enabledIDs[f.ID] = true
	}

	// Bucket features by category, preserving registry order.
	buckets := make(map[feature.Category][]*feature.Feature)
	for _, f := range ctx.EnabledFeatures {
		buckets[f.Category] = append(buckets[f.Category], f)
	}

	var groups []BuildArgGroup
	for _, cat := range categoryOrder {
		features := buckets[cat]
		if len(features) == 0 {
			continue
		}

		var entries []BuildArgEntry
		// Track base features that are suppressed because their dev counterpart is present.
		suppressed := make(map[string]bool)
		for _, f := range features {
			if f.IsDev && f.BaseLang != "" && enabledIDs[f.BaseLang] {
				suppressed[f.BaseLang] = true
			}
		}

		for _, f := range features {
			// Skip base feature if its dev counterpart is also enabled.
			if suppressed[f.ID] {
				continue
			}

			entry := BuildArgEntry{Line: f.BuildArg + "=true"}
			if f.IsDev && f.BaseLang != "" {
				// Look up the base feature's display name for the comment.
				for _, base := range features {
					if base.ID == f.BaseLang {
						entry.Comment = "includes " + base.DisplayName + " runtime"
						break
					}
				}
			}
			entries = append(entries, entry)

			// Attach version arg: for dev features, use the base feature's version.
			versionFeature := f
			if f.IsDev && f.BaseLang != "" {
				for _, base := range features {
					if base.ID == f.BaseLang {
						versionFeature = base
						break
					}
				}
			}
			if versionFeature.VersionArg != "" {
				ver := ctx.Versions[versionFeature.VersionArg]
				if ver != "" {
					entries = append(entries, BuildArgEntry{
						Line: versionFeature.VersionArg + "=" + ver,
					})
				}
			}
		}

		if len(entries) > 0 {
			groups = append(groups, BuildArgGroup{
				Label: categoryLabel[cat],
				Args:  entries,
			})
		}
	}

	return groups
}

func funcMap() template.FuncMap {
	return template.FuncMap{
		"hasFeature": func(ctx *RenderContext, id string) bool {
			return ctx.Selection.Has(id)
		},
		"indent": func(n int, s string) string {
			pad := strings.Repeat(" ", n)
			lines := strings.Split(s, "\n")
			for i, line := range lines {
				if line != "" {
					lines[i] = pad + line
				}
			}
			return strings.Join(lines, "\n")
		},
		"joinComma": func(items []string) string {
			return strings.Join(items, ", ")
		},
		"buildArgLines": func(ctx *RenderContext) string {
			var lines []string
			for _, f := range ctx.EnabledFeatures {
				lines = append(lines, "- "+f.BuildArg+"=true")
				if f.VersionArg != "" {
					ver := ctx.Versions[f.VersionArg]
					if ver == "" {
						ver = f.DefaultVersion
					}
					lines = append(lines, "- "+f.VersionArg+"="+ver)
				}
			}
			return strings.Join(lines, "\n")
		},
		"groupedBuildArgs": func(ctx *RenderContext) []BuildArgGroup {
			return groupedBuildArgs(ctx)
		},
		"add": func(a, b int) int {
			return a + b
		},
		"sub": func(a, b int) int {
			return a - b
		},
		"contains": func(s, substr string) bool {
			return strings.Contains(s, substr)
		},
		"split": func(s, sep string) []string {
			return strings.Split(s, sep)
		},
	}
}

File: crates/containers-common/src/template/funcmap.rs
//! Template functions — `groupedBuildArgs` and related types.

use crate::feature::{Category, Feature};

use super::context::RenderContext;

/// A category of build args with a label header.
#[derive(Debug, Clone)]
pub struct BuildArgGroup {
    /// Display label (e.g. "Languages", "Tools").
    pub label: String,

    /// Ordered entries within the group.
    pub args: Vec<BuildArgEntry>,
}

/// A single build arg line with an optional inline comment.
#[derive(Debug, Clone)]
pub struct BuildArgEntry {
    /// e.g. `"INCLUDE_PYTHON_DEV=true"`.
    pub line: String,

    /// e.g. `"includes Python runtime"` (empty if none).
    pub comment: String,
}

/// Maps feature categories to display labels.
const fn category_label(cat: Category) -> &'static str {
    match cat {
        Category::Language => "Languages",
        Category::Tool => "Tools",
        Category::Ai => "AI/ML",
        Category::Cloud => "Cloud",
        Category::Database => "Database",
    }
}

/// Category rendering order.
const CATEGORY_ORDER: [Category; 5] =
    [Category::Language, Category::Tool, Category::Ai, Category::Cloud, Category::Database];

/// Groups enabled features by category, collapses base+dev pairs (emitting
/// only the dev line with an annotation), and attaches version args after the
/// corresponding feature entry.
#[must_use]
pub fn grouped_build_args(ctx: &RenderContext) -> Vec<BuildArgGroup> {
    // Build a set of enabled feature IDs for quick lookup.
    let enabled_ids: std::collections::HashSet<&str> =
        ctx.enabled_features.iter().map(|f| f.id.as_str()).collect();

    // Bucket features by category, preserving registry order.
    let mut buckets: std::collections::HashMap<Category, Vec<&Feature>> =
        std::collections::HashMap::new();
    for f in &ctx.enabled_features {
        buckets.entry(f.category).or_default().push(f);
    }

    let mut groups = Vec::new();
    for cat in CATEGORY_ORDER {
        let Some(features) = buckets.get(&cat) else {
            continue;
        };
        if features.is_empty() {
            continue;
        }

        // Track base features suppressed because their dev counterpart is present.
        let mut suppressed = std::collections::HashSet::new();
        for f in features {
            if f.is_dev
                && let Some(base_lang) = &f.base_lang
                && enabled_ids.contains(base_lang.as_str())
            {
                suppressed.insert(base_lang.as_str());
            }
        }

        let mut entries = Vec::new();
        for f in features {
            // Skip base feature if its dev counterpart is also enabled.
            if suppressed.contains(f.id.as_str()) {
                continue;
            }

            let mut entry =
                BuildArgEntry { line: format!("{}=true", f.build_arg), comment: String::new() };

            // For dev features, add "includes X runtime" comment.
            if f.is_dev
                && let Some(base_lang) = &f.base_lang
            {
                for base in features {
                    if base.id == *base_lang {
                        entry.comment = format!("includes {} runtime", base.display_name);
                        break;
                    }
                }
            }
            entries.push(entry);

            // Attach version arg: for dev features, use the base feature's version.
            let version_feature = if f.is_dev {
                f.base_lang
                    .as_ref()
                    .and_then(|bl| features.iter().find(|base| base.id == *bl))
                    .unwrap_or(f)
            } else {
                f
            };

            if let Some(version_arg) = &version_feature.version_arg
                && let Some(ver) = ctx.versions.get(version_arg)
                && !ver.is_empty()
            {
                entries.push(BuildArgEntry {
                    line: format!("{version_arg}={ver}"),
                    comment: String::new(),
                });
            }
        }

        if !entries.is_empty() {
            groups.push(BuildArgGroup { label: category_label(cat).into(), args: entries });
        }
    }

    groups
}
